swupd/delta.go

```diff
@@ -0,0 +1,190 @@
+		return nil, err
+	}
+
+	// Must be sorted before passing into linkPeersAndChange()
+	oldManifest.sortFilesName()
+	newManifest.sortFilesName()
+
+	_, _, _ = newManifest.linkPeersAndChange(oldManifest, from)
+
+	deltas, err := consolidateDeltaFiles(newManifest, from, c)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to create deltas list %s", manifest)
+	}
+
+	const maxRoutines = 3
+	var failed []*File
+	var deltaQueue = make(chan *File)
+	var failedChan = make(chan *File, maxRoutines)
+	var wg sync.WaitGroup
+	wg.Add(maxRoutines)
+
+	for i := 0; i < maxRoutines; i++ {
+		go func(statedir string) {
+			defer wg.Done()
+
+			for f := range deltaQueue {
+				err := createDelta(f, statedir, f.DeltaPeer.Version, f.DeltaPeer.Hash)
+				if err != nil {
+					failedChan <- f
+				}
+			}
+		}(statedir)
+	}
+
+	for _, file := range deltas {
+		select {
+		case deltaQueue <- file:
+		case f := <-failedChan:
+			failed = append(failed, f)
+		}
+	}
+
+	close(deltaQueue)
+	wg.Wait()
+
+	if len(failedChan) > 0 {
+		f := <-failedChan
+		failed = append(failed, f)
+	}
+
+	return failed, nil
+}
+
+func consolidateDeltaFiles(manifest *Manifest, from uint32, c config) ([]*File, error) {
+	var files []*File
+
+	if manifest == nil {
+		return nil, nil
+	}
+
+	for _, file := range manifest.Files {
+		if file.Version <= from {
+			continue
+		}
+		if file.DeltaPeer == nil {
+			continue
+		}
+		if file.Type != TypeFile ||
+			file.DeltaPeer.Type != TypeFile {
+			continue
+		}
+
+		deltadir := filepath.Join(c.outputDir, fmt.Sprintf("%d", file.Version), "delta")
+		fromToString := fmt.Sprintf("%d-%d-%s-%s", file.DeltaPeer.Version, file.Version, file.DeltaPeer.Hash, file.Hash)
+
+		deltaFile := filepath.Join(deltadir, fromToString)
+
+		// Only add to list if the delta file does not already exist on the system
+		if _, err := os.Stat(deltaFile); os.IsNotExist(err) {
+			// Only create deltas for files bigger than 200 bytes
+			fullname := filepath.Join(c.imageBase, fmt.Sprintf("%d", file.Version), "full"+file.Name)
+			file.Info, _ = os.Stat(fullname)
+			if file.Info.Size() > 200 {
+				files = append(files, file)
+			}
+		}
+
+	}
+
+	return files, nil
+}
+
+func createDelta(file *File, statedir string, fromVersion uint32, fromHash Hashval) error {
+	var c config
+	c, err := getConfig(statedir)
+	if err != nil {
+		return err
+	}
+
+	// File without peers cannot have deltas
+	if file.DeltaPeer == nil {
+		return nil
+	}
+
+	// We only support deltas between two regular files for now
+	// TODO: Support directory -> file in the future, this is a complex case
+	if file.Type != TypeFile || file.DeltaPeer.Type != TypeFile {
+		return nil
+	}
+
+	// By definition we cannot make deltas for files that are deleted
+	if file.Status == StatusDeleted {
+		return nil
+	}
+
+	// To files cannot be ghosted
+	if file.Status == StatusGhosted {
+		return nil
+	}
```

Comment:
Given this is a helper function for creating deltas we already identified (so, not really meant to be used in isolation), ideally we shouldn't need to perform those checks again.

If you must perform those checks, move the checks to a function and call the same checks from here and consolidateDeltas.